osu.Game/Beatmaps/BeatmapDifficultyManager.cs

```diff
@@ -109,43 +109,40 @@ public StarDifficulty GetDifficulty([NotNull] BeatmapInfo beatmapInfo, [CanBeNul
         }
 
         private CancellationTokenSource trackedUpdateCancellationSource;
+        private readonly List<CancellationTokenSource> linkedCancellationSources = new List<CancellationTokenSource>();
 
         /// <summary>
         /// Updates all tracked <see cref="BindableStarDifficulty"/> using the current ruleset and mods.
         /// </summary>
         private void updateTrackedBindables()
         {
-            trackedUpdateCancellationSource?.Cancel();
+            cancelTrackedBindableUpdate();
             trackedUpdateCancellationSource = new CancellationTokenSource();
 
             foreach (var b in trackedBindables)
             {
-                if (trackedUpdateCancellationSource.IsCancellationRequested)
-                    break;
+                var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(trackedUpdateCancellationSource.Token, b.CancellationToken);
+                linkedCancellationSources.Add(linkedSource);
 
-                using (var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(trackedUpdateCancellationSource.Token, b.CancellationToken))
-                    updateBindable(b, currentRuleset.Value, currentMods.Value, linkedSource.Token);
+                updateBindable(b, currentRuleset.Value, currentMods.Value, linkedSource.Token);
             }
         }
 
         /// <summary>
-        /// Updates the value of a <see cref="BindableStarDifficulty"/> with a given ruleset + mods.
+        /// Cancels the existing update of all tracked <see cref="BindableStarDifficulty"/> via <see cref="updateTrackedBindables"/>.
         /// </summary>
-        /// <param name="bindable">The <see cref="BindableStarDifficulty"/> to update.</param>
-        /// <param name="rulesetInfo">The <see cref="RulesetInfo"/> to update with.</param>
-        /// <param name="mods">The <see cref="Mod"/>s to update with.</param>
-        /// <param name="cancellationToken">A token that may be used to cancel this update.</param>
-        private void updateBindable([NotNull] BindableStarDifficulty bindable, [CanBeNull] RulesetInfo rulesetInfo, [CanBeNull] IEnumerable<Mod> mods, CancellationToken cancellationToken = default)
+        private void cancelTrackedBindableUpdate()
         {
-            GetDifficultyAsync(bindable.Beatmap, rulesetInfo, mods, cancellationToken).ContinueWith(t =>
+            trackedUpdateCancellationSource?.Cancel();
+            trackedUpdateCancellationSource = null;
+
+            foreach (var c in linkedCancellationSources)
             {
-                // We're on a threadpool thread, but we should exit back to the update thread so consumers can safely handle value-changed events.
-                Schedule(() =>
-                {
-                    if (!cancellationToken.IsCancellationRequested)
-                        bindable.Value = t.Result;
-                });
-            }, cancellationToken);
+                c.Cancel();
```

Comment:
Not sure what this second cancel is for - shouldn't all `linkedCancellationSources` get cancelled once `trackedUpdateCancellationSource` is?

Reply:
Sure, it was mostly for safeguarding.